src/parser/jobs/mnk/BuffUptime.js

```diff
@@ -0,0 +1,153 @@
+import React, {Fragment} from 'react'
+
+import {ActionLink} from 'components/ui/DbLink'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {Rule, Requirement} from 'parser/core/modules/Checklist'
+import {Suggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+// in ms
+const DRAGON_KICK_DURATION = 15000
+const TWIN_SNAKES_DURATION = 15000
+const THREE_GCD_BUFFER = 6021 // TODO: get this from gcd??
+
+export default class BuffUptime extends Module {
+	static handle = 'BuffUptime'
+	static dependencies = [
+		'checklist',
+		'combatants',
+		'enemies',
+		'invuln',
+		'suggestions',
+	]
+
+	// Dragon kick can be applied to multiple enemies
+	_lastDragonKickUse = {}
+	// Twin snakes can only be applied to self
+	_lastTwinSnakesUse = {}
+
+	_earlyDragonKicks = 0
+	_earlyTwinSnakes = 0
+
+	constructor(...args) {
+		super(...args)
+
+		const dk_filter = {
+			by: 'player',
+			abilityId: ACTIONS.DRAGON_KICK.id,
+		}
+
+		const ts_filter = {
+			by: 'player',
+			abilityId: STATUSES.TWIN_SNAKES.id,
+		}
+
+		// Using cast instead of applydebuff because the blunt resistance down debuff
+		// gets applied in the logs much more often than dragon kick is actually
+		// cast
+		this.addHook('cast', dk_filter, this._onDragonKickApplication)
+		this.addHook(['applybuff', 'refreshbuff'], ts_filter, this._onTwinSnakesApplication)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onDragonKickApplication(event) {
+		// Make sure we're tracking for this target
+		const lastApplication = this._lastDragonKickUse[event.targetID]
+
+		// If it's not been applied yet set it and skip out
+		if (!lastApplication) {
+			this._lastDragonKickUse[event.targetID] = event.timestamp
+			return
+		}
+
+		const timeSinceLastApplication = event.timestamp - lastApplication
+		if (timeSinceLastApplication < DRAGON_KICK_DURATION - THREE_GCD_BUFFER) {
+			this._earlyDragonKicks++
+		}
+		this._lastDragonKickUse[event.targetID] = event.timestamp
+	}
+
+	_onTwinSnakesApplication(event) {
+		// If it's not been applied yet set it and skip out
+		if (!self._lastTwinSnakesUse) {
+			self._lastTwinSnakesUse = event.timestamp
+			return
+		}
+
+		const timeSinceLastApplication = event.timestamp - self._lastTwinSnakesUse
+		if (timeSinceLastApplication < TWIN_SNAKES_DURATION - THREE_GCD_BUFFER) {
```

Comment:
The issue here is the very poor GCD estimate right? Without a knowing the GCD time of the player we can't know for sure if they should reapply TS or DK since we don't know if they will have another chance before it drops. I'm not sure the onCast approach solves that without duplicating what a GCD module would do.  I'm inclined to leave it as is and update it to use the GCD module once it works properly for monk. As is, it still provides value, I can see obvious differences between 50% and 100% MNK logs. Thoughts?

Reply:
GCD stuff has actually changed recently, but I don't think in a way we can use for this. What are you actually trying to calculate from this check? I can read the code, but I want to know what you think you're trying to get out of it. I've done a tonne of work the last few days on GL stuff and moving onto buffs and whatnot so maybe you can hit me up on Discord (acchan#5683) and we can work over this in real time.